syntax = "proto3";

// Internal
service Node {
  rpc AppendEntries (AppendRequest) returns (AppendResponse) {}
  rpc RequestVote (VoteRequest) returns (VoteResponse) {}
}

message AppendRequest {
    int32 term = 1;
    int32 leaderID = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated int32 entries = 5;
    int32 leaderCommitIndex = 6;
}
message AppendResponse {
    int32 term = 1;
    int32 ackIndex = 2;
    int32 nodeID = 4;
    bool success = 3;    
}

message VoteRequest {
    int32 term = 1;
    int32 candidateId = 2;  
    int32 lastLogIndex  = 3;
    int32 lastLogTerm = 4;
}
message VoteResponse {
    int32 resID = 3;
    int32 term = 1;
    bool voteGranted = 2;
}

// Client side
service Client {
    rpc ServeClient (ServeClientArgs) returns (ServeClientReply) {}
}

message ServeClientArgs {
    string request = 1;
}
  
message ServeClientReply {
    string data = 1;
    string leaderID = 2;
    bool success = 3;
}
